match/lib/match/encryption/encryption.rb

```diff
@@ -0,0 +1,150 @@
+
+      def encrypt(data, password, salt)
+        cipher = ::OpenSSL::Cipher.new(ALGORITHM)
+        cipher.encrypt
+
+        keyivgen(cipher, password, salt)
+
+        encrypted_data = cipher.update(data)
+        encrypted_data << cipher.final
+
+        auth_tag = cipher.auth_tag
+
+        { encrypted_data: encrypted_data, auth_tag: auth_tag }
+      end
+
+      def decrypt(encrypted_data, password, salt, auth_tag)
+        cipher = ::OpenSSL::Cipher.new(ALGORITHM)
+        cipher.decrypt
+
+        keyivgen(cipher, password, salt)
+
+        cipher.auth_tag = auth_tag
+
+        data = cipher.update(encrypted_data)
+        data << cipher.final
+      end
+
+      private
+
+      def keyivgen(cipher, password, salt)
+        keyIv = ::OpenSSL::KDF.pbkdf2_hmac(password, salt: salt, iterations: 10_000, length: 32 + 12 + 24, hash: "sha256")
+        key = keyIv[0..31]
+        iv = keyIv[32..43]
+        auth_data = keyIv[44..-1]
+        cipher.key = key
+        cipher.iv = iv
+        cipher.auth_data = auth_data
+      end
+    end
+
+    class MatchDataEncryption
+      V1_PREFIX = "Salted__"
+      V2_PREFIX = "match_encrypted_v2__"
+
+      def encrypt(data, password, version = 2)
+        salt = SecureRandom.random_bytes(8)
+        if version == 2
+          e = EncryptionV2.new
+          encryption = e.encrypt(data, password, salt)
+          encrypted_data = V2_PREFIX + salt + encryption[:auth_tag] + encryption[:encrypted_data]
+        else
+          e = EncryptionV1.new
+          encryption = e.encrypt(data, password, salt)
+          encrypted_data = V1_PREFIX + salt + encryption[:encrypted_data]
+        end
+        Base64.encode64(encrypted_data)
+      end
+
+      def decrypt(base64encoded_encrypted, password)
+        stored_data = Base64.decode64(base64encoded_encrypted)
+        if stored_data.start_with?(V2_PREFIX)
+          salt = stored_data[20..27]
+          auth_tag = stored_data[28..43]
+          data_to_decrypt = stored_data[44..-1]
+          e = EncryptionV2.new
+          e.decrypt(data_to_decrypt, password, salt, auth_tag)
+        else
+          salt = stored_data[8..15]
+          data_to_decrypt = stored_data[16..-1]
+          e = EncryptionV1.new
+          begin
+            e.decrypt(data_to_decrypt, password, salt)
+          rescue => _ex
+            # Note that we are not guaranteed to catch the decryption errors here if the password is wrong
+            # as there's no integrity checks.
+            # With a wrong password, there's a 0.4% chance it will decrypt garbage and not fail
+            # see https://github.com/fastlane/fastlane/issues/21663
+            fallback_hash_algorithm = "SHA256"
+            e.decrypt(data_to_decrypt, password, salt, fallback_hash_algorithm)
+          end
+        end
+      end
+    end
+
+    # by default, encrypt or decrypt in place
+    class MatchFileEncryption
+      def encrypt(file_path, password, output_path = nil)
```

Comment:
Another question I have: do we want to enforce keywords parameters instead?

I think it usually makes for better API management and backwards compatibility.

Should I update this file and use keywords args instead?

Reply:
I added some but not in all classes. Makes sense or inconsistent?